### PR TITLE
Integrate Kepler.gl as a new visualization type

### PR DESCRIPTION
**Pull request by mistercrunch**
Kepler.gl is an amazing geospatial React application. It offers a very
well designed user interface to compose geo layers and slice and dice
geospatial data.

screen shot 2018-11-16 at 12 12 20 am

As it turns out, Kepler is designed to be embedded as part of React
apps. It took me only a few hours to get to a working prototype.

Next steps:
<strike>
* allow to save the state, without the data
* restore the state, and merge with the new data
</strike>
* tweak styles, make a light style that looks decent within Superset,
  Kepler uses stylable components, so that should be possible

### CODE REVIEW

**Comment by mistercrunch** [comment]
Improved the solution to support saving the state properly, using its own redux store, autozoom, readonly and more.

File: superset/assets/package.json [hunk 1]
```diff
@@ -81,7 +81,7 @@
     "d3-tip": "^0.9.1",
     "datamaps": "^0.5.8",
     "datatables.net-bs": "^1.10.15",
-    "deck.gl": "^5.3.5",
+    "deck.gl": "^5.3.4",
```
Comment (mistercrunch) [inline superset/assets/package.json]: This is to align Kepler's deck.gl dep with Superset's. It's important as having 2 versions of deck.gl errors out.

**Comment by heshan0131** [comment]
Accidentally found this PR. Great work! let us know if you need any help on our side. I found your next item is to change the styling. You can follow [This issue][https://github.com/uber/kepler.gl/issues/374], it will make overriding style a lot easier!

**Comment by vylc** [comment]
@mistercrunch any movement on this in getting to the next version of deck?

**Comment by mistercrunch** [comment]
@heshan0131 I missed the notification on your previous comment, but thank you for your note! The main things I had questions around are:

* are there any events I could hook onto when controls are changing? for now I had to hook up something to the `onSaveMap` event. Superset assumes that the visualization controls are outside the visualization, but passes a method to the visualization to allow it to change the value of a superset control. For now what I do is dump the state of Kepler inside a Superset TextControl, but we had to tell the users to go and push that button prior to saving... I wish I have a generic `onKeplerStateChange` prop or something like that...
* is there a light style mode as opposed to the dark theme? I dug through the Kepler code and found some reference to a light theme, but it seemed to apply only to some modals... Maybe we just need a dark mode for Superset :)

**Comment by heshan0131** [comment]
@mistercrunch kepler.gl currently doesn't have a prop to listen on `onStateChange` event, but you are right, we should provide more callbacks such as `onStateChange` `onMouseHover` etc. We are slowly learning how people use the framework. Before we add that, you can add middleware to your store and use `KeplerGlSchema` to serialize the state then diff it.

Something like:

```js
import {KeplerGlSchema} from 'kepler.gl/schemas';
import {taskMiddleware} from 'react-palm/tasks';

let previousValue = null;

const stateChangeMiddleware = store => next => action => {
  const returnValue = next(action);

  const keplerState = store.getState().keplerGl.map;
  const stateToSave = KeplerGlSchema.getConfigToSave(keplerState);
  const hash = JSON.stringify(stateToSave);
    
  if (previousValue !== hash) {
     // do something here
     console.log('kepler.gl state has changed');
     previousValue = hash;
   }

    return returnValue;
};

const middlewares =[taskMiddleware, stateChangeMiddleware];
const enhancers = [applyMiddleware(...middlewares)];
const store = createStore(reducers, {}, compose(...enhancers));

return store;
```

**Comment by heshan0131** [comment]
Also we don't have a complete light theme yet (also on our todo list)
@macrigiuseppe has created a [custom theme example](https://github.com/uber/kepler.gl/blob/master/examples/custom-theme/src/app.js#L29-L48) which contains a fairly reasonable light theme, just need a little bit touch up. Worth a try

**Comment by mistercrunch** [comment]
@heshan0131 this is super helpful, thanks for the input and the great tool!

**Comment by mistercrunch** [comment]
@heshan0131, I made it work! I had to adapt the code a little and rate limit the changes as it made the map interactions sluggish (we update a control onChange and causes a chain of renders).

**Comment by mistercrunch** [comment]
Looks like theme support is in a newer version. I have a branch trying to upgrade kepler + deck to latest

**Comment by mistercrunch** [comment]
Tried theming kepler 1.0 (not released yet) and it looks promising but didn't work out of the box
Screen Shot 2019-04-09 at 10 54 24 PM
Screen Shot 2019-04-09 at 10 54 12 PM

**Comment by heshan0131** [comment]
> @heshan0131, I made it work! I had to adapt the code a little and rate limit the changes as it made the map interactions sluggish (we update a control onChange and causes a chain of renders).


You can try skip [`ActionTypes.LAYER_HOVER`](https://github.com/uber/kepler.gl/blob/master/docs/api-reference/actions/actions.md#onlayerhover) and [`ActionTypes.UPDATE_MAP`](https://github.com/uber/kepler.gl/blob/master/docs/api-reference/actions/actions.md#updatemap). This will significantly reduce state serialization being called

using `import {ActionTypes} from 'kepler.gl/actions';` to get access to the constant

**Comment by mistercrunch** [comment]
Good input, thought we want to capture the viewport in the state, so the rate limiting option seems like like the best option for now.

**Comment by mistercrunch** [comment]
@betodealmeida @xtinec @vylc Kepler properly save its state into Superset after the changes I added recently

**Comment by chintanp** [comment]
@mistercrunch Any ideas when this will make it to Superset?

**Comment by mistercrunch** [comment]
It should be easy to re-do this based on the new plugin interface.
https://preset.io/blog/2020-07-02-hello-world/

**Comment by DrissiReda** [comment]
@mistercrunch I still had issues with that. I tried using both kepler and deckgl  then I tried replacing deckgl with kepler, these are the errors I get:

- When replaced : ` "export 'DeckGLChartPreset' was not found in '@superset-ui/legacy-preset-chart-kepler' `
- When using both: `"export 'KeplerChartPreset' was not found in '@superset-ui/legacy-preset-chart-kepler'`
- When using both: `"export 'KeplerChartPlugin' was not found in '@superset-ui/legacy-preset-chart-kepler'`
- When using both: `"export 'Kepler' was not found in '@superset-ui/legacy-preset-chart-kepler'`

What exactly should I use, the README.md of the kepler plugin is a copy paste of the deckgl plugin.

**Pull request closed without merging**